cmd/kubeadm/app/componentconfigs/configset.go

```diff
@@ -0,0 +1,204 @@
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	"k8s.io/kubernetes/cmd/kubeadm/app/util/apiclient"
+)
+
+// handler is a package internal type that handles component config factory and common functionality.
+// Every component config group should have exactly one static instance of handler.
+type handler struct {
+	// GroupVersion holds this handler's group name and preferred version
+	GroupVersion schema.GroupVersion
+
+	// AddToScheme points to a func that should add the GV types to a schema
+	AddToScheme func(*runtime.Scheme) error
+
+	// CreateEmpty returns an empty kubeadmapi.ComponentConfig (not even defaulted)
+	CreateEmpty func() kubeadmapi.ComponentConfig
+
+	// fromCluster should load the component config from a config map on the cluster.
+	// Don't use this directly! Use FromCluster instead!
+	fromCluster func(*handler, clientset.Interface, *kubeadmapi.ClusterConfiguration) (kubeadmapi.ComponentConfig, error)
+}
+
+// FromDocumentMap looks in the document map for documents with this handler's group.
+// If such are found a new component config is instantiated and the documents are loaded into it.
+// No error is returned if no documents are found.
+func (h *handler) FromDocumentMap(docmap kubeadmapi.DocumentMap) (kubeadmapi.ComponentConfig, error) {
+	for gvk := range docmap {
+		if gvk.Group == h.GroupVersion.Group {
+			cfg := h.CreateEmpty()
+			if err := cfg.Unmarshal(docmap); err != nil {
+				return nil, err
+			}
+			return cfg, nil
+		}
+	}
+	return nil, nil
+}
+
+// fromConfigMap is an utility function, which will load the value of a key of a config map and use h.FromDocumentMap() to perform the parsing
+// This is an utility func. Used by the component config support implementations. Don't use it outside of that context.
+func (h *handler) fromConfigMap(client clientset.Interface, cmName, cmKey string, mustExist bool) (kubeadmapi.ComponentConfig, error) {
+	configMap, err := apiclient.GetConfigMapWithRetry(client, metav1.NamespaceSystem, cmName)
+	if err != nil {
+		if !mustExist && (apierrors.IsNotFound(err) || apierrors.IsForbidden(err)) {
+			klog.Warningf("Warning: No %s config is loaded. Continuing without it: %v", h.GroupVersion, err)
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	configData, ok := configMap.Data[cmKey]
+	if !ok {
+		return nil, errors.Errorf("unexpected error when reading %s ConfigMap: %s key value pair missing", cmName, cmKey)
+	}
+
+	gvkmap, err := kubeadmutil.SplitYAMLDocuments([]byte(configData))
+	if err != nil {
+		return nil, err
+	}
+
+	return h.FromDocumentMap(gvkmap)
+}
+
+// FromCluster loads a component from a config map in the cluster
+func (h *handler) FromCluster(clientset clientset.Interface, clusterCfg *kubeadmapi.ClusterConfiguration) (kubeadmapi.ComponentConfig, error) {
+	return h.fromCluster(h, clientset, clusterCfg)
```

Comment:
Can't we make `fromCluster` a `handler` method as well, so there's no need to pass `h` as argument?

Reply:
I see why you did this on the non-public one... nevermind.